dataretrieval/nldi.py

```diff
@@ -0,0 +1,495 @@
+                    'navigation_mode is required if comid or data_source is provided'
+                )
+        # validate the feature source and comid
+        _validate_feature_source_comid(feature_source, feature_id, comid)
+        # validate the data source
+        if data_source:
+            _validate_data_source(data_source)
+        # validate feature source
+        _validate_data_source(feature_source)
+        # validate the navigation mode
+        if navigation_mode:
+            _validate_navigation_mode(navigation_mode)
+
+    if lat:
+        url = f'{NLDI_API_BASE_URL}/comid/position'
+        query_params = {'coords': f'POINT({long} {lat})'}
+    else:
+        if navigation_mode:
+            if feature_source:
+                url = f'{NLDI_API_BASE_URL}/{feature_source}/{feature_id}/navigation'
+            else:
+                url = f'{NLDI_API_BASE_URL}/comid/{comid}/navigation'
+            url += f'/{navigation_mode}/{data_source}'
+            query_params = {'distance': str(distance)}
+            if stop_comid is not None:
+                query_params['stopComid'] = str(stop_comid)
+        else:
+            url = f'{NLDI_API_BASE_URL}/{feature_source}/{feature_id}'
+            query_params = {}
+
+    if lat:
+        err_msg = f"Error getting features for lat '{lat}'" f" and long '{long}'"
+    elif feature_source:
+        err_msg = (
+            f"Error getting features for feature source '{feature_source}'"
+            f" and feature_id '{feature_id}, and data source '{data_source}'"
+        )
+    else:
+        err_msg = (
+            f"Error getting features for comid '{comid}'"
+            f" and data source '{data_source}'"
+        )
+
+    feature_collection = _query_nldi(url, query_params, err_msg)
+    if as_json:
+        return feature_collection
+    gdf = gpd.GeoDataFrame.from_features(feature_collection)
+    return gdf
+
+
+# TODO: This function can cause timeout error for some data sources
+#  - may be we shouldn't provide this function?
+def get_features_by_data_source(data_source: str) -> gpd.GeoDataFrame:
+    """Gets all features found for the specified data source as
+    points in WGS84 lat/long coordinates as GeoDataFrame containing a point geometry.
+
+    Parameters
+    ----------
+    data_source: string, data source
+
+    Returns
+    -------
+    gdf: GeoDataFrame
+        GeoDataFrame of features
+
+    Examples
+    --------
+    .. doctest::
+
+        >>> # Get features for a feature wqp and feature_id USGS-01031500
+        >>> gdf = dataretrieval.nldi.get_features_by_data_source(data_source="nwissite")
+    """
+    # validate the data source
+    _validate_data_source(data_source)
+    url = f'{NLDI_API_BASE_URL}/{data_source}'
+    err_msg = f"Error getting features for data source '{data_source}'"
+    feature_collection = _query_nldi(url, {}, err_msg)
+    gdf = gpd.GeoDataFrame.from_features(feature_collection)
+    return gdf
+
+
+def search(
```

Comment:
I intend search to be a lower-level function to return a nested dict of multiple feature types. For example, if we wanted to get multiple features upstream of a site, this would allow us to accomplish that in a single API call. Nevertheless, I'm fine with this for the first cut. Great work!

Reply:
Due to my lack of understanding of the domain, I was struggling a bit to implement the search functionality.